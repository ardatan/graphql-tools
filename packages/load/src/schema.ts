import {
  BuildSchemaOptions,
  GraphQLSchema,
  Source as GraphQLSource,
  lexicographicSortSchema,
  print,
} from 'graphql';
import {
  extractExtensionsFromSchema,
  IExecutableSchemaDefinition,
  mergeSchemas,
} from '@graphql-tools/schema';
import {
  getResolversFromSchema,
  IResolvers,
  SchemaExtensions,
  Source,
  TypeSource,
} from '@graphql-tools/utils';
import { OPERATION_KINDS } from './documents.js';
import {
  loadTypedefs,
  LoadTypedefsOptions,
  loadTypedefsSync,
  UnnormalizedTypeDefPointer,
} from './load-typedefs.js';

export type LoadSchemaOptions = BuildSchemaOptions &
  LoadTypedefsOptions &
  Partial<IExecutableSchemaDefinition> & {
    /**
     * Adds a list of Sources in to `extensions.sources`
     *
     * Disabled by default.
     */
    includeSources?: boolean;
  };

/**
 * Asynchronously loads a schema from the provided pointers.
 * @param schemaPointers Pointers to the sources to load the schema from
 * @param options Additional options
 */
export async function loadSchema(
  schemaPointers: UnnormalizedTypeDefPointer | UnnormalizedTypeDefPointer[],
  options: LoadSchemaOptions,
): Promise<GraphQLSchema> {
  const sources = await loadTypedefs(schemaPointers, {
    ...options,
    filterKinds: OPERATION_KINDS,
  });
  return getSchemaFromSources(sources, options);
}

/**
 * Synchronously loads a schema from the provided pointers.
 * @param schemaPointers Pointers to the sources to load the schema from
 * @param options Additional options
 */
export function loadSchemaSync(
  schemaPointers: UnnormalizedTypeDefPointer | UnnormalizedTypeDefPointer[],
  options: LoadSchemaOptions,
): GraphQLSchema {
  const sources = loadTypedefsSync(schemaPointers, {
    filterKinds: OPERATION_KINDS,
    ...options,
  });
  return getSchemaFromSources(sources, options);
}

function includeSources(schema: GraphQLSchema, sources: Source[]) {
  const finalSources: Array<GraphQLSource> = [];
  for (const source of sources) {
    if (source.rawSDL) {
      finalSources.push(new GraphQLSource(source.rawSDL, source.location));
    } else if (source.document) {
      finalSources.push(new GraphQLSource(print(source.document), source.location));
    }
  }
  schema.extensions = {
    ...schema.extensions,
    sources: finalSources,
    extendedSources: sources,
  };
}

function getSchemaFromSources(sources: Source[], options: LoadSchemaOptions) {
  if (
    sources.length === 1 &&
    sources[0].schema != null &&
    options.typeDefs == null &&
    options.resolvers == null
  ) {
    return options.sort ? lexicographicSortSchema(sources[0].schema) : sources[0].schema;
  }
  const { typeDefs, resolvers, schemaExtensions } = collectSchemaParts(sources);

  const schema = mergeSchemas({
    ...options,
    typeDefs,
    resolvers,
    schemaExtensions,
  });

  if (options?.includeSources) {
    includeSources(schema, sources);
  }

  return options.sort ? lexicographicSortSchema(schema) : schema;
}

function collectSchemaParts(sources: Source[]) {
  const typeDefs: TypeSource[] = [];
  const resolvers: IResolvers[] = [];
  const schemaExtensions: SchemaExtensions[] = [];

  for (const source of sources) {
    if (source.schema) {
      typeDefs.push(source.schema);
      resolvers.push(getResolversFromSchema(source.schema));
      schemaExtensions.push(extractExtensionsFromSchema(source.schema));
    } else {
      const typeDef = source.document || source.rawSDL;
      if (typeDef) {
        typeDefs.push(typeDef);
      }
    }
  }

  return {
    typeDefs,
    resolvers,
    schemaExtensions,
  };
}
