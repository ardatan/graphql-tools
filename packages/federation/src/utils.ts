import { GraphQLSchema } from 'graphql';
import { FilterRootFields, FilterTypes } from '@graphql-tools/wrap';

export function getArgsFromKeysForFederation(representations: readonly any[]) {
  return { representations };
}

export function getKeyForFederation(root: any) {
  return root;
}

export function filterInternalFieldsAndTypes(finalSchema: GraphQLSchema) {
  const removeEntitiesField = new FilterRootFields(
    (operation, fieldName) =>
      !(operation === 'Query' && (fieldName === '_entities' || fieldName === '_sdl')),
  );
  const removeEntityAndAny = new FilterTypes(
    type =>
      type.name !== '_Entity' &&
      type.name !== '_Any' &&
      type.name !== '_FieldSet' &&
      type.name !== '_Service' &&
      !type.name.startsWith('link__'),
  );
  const fakeSubschemaConfig = {
    schema: finalSchema,
    transforms: [removeEntitiesField, removeEntityAndAny],
  };
  for (const transform of fakeSubschemaConfig.transforms) {
    finalSchema = transform.transformSchema(finalSchema, fakeSubschemaConfig);
  }
  return finalSchema;
}
