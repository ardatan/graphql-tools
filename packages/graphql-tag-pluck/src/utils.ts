// Will use the shortest indention as an axis
export const freeText = (text: string | string[], skipIndentation = false) => {
  if (text instanceof Array) {
    text = text.join('');
  }

  // This will allow inline text generation with external functions, same as ctrl+shift+c
  // As long as we surround the inline text with ==>text<==
  text = text.replace(/( *)==>((?:.|\n)*?)<==/g, (_match, baseIndent, content: string) => {
    return content
      .split('\n')
      .map(line => `${baseIndent}${line}`)
      .join('\n');
  });

  if (skipIndentation) {
    if (text.trim() === '') {
      return '';
    }
    return text;
  }

  const lines = text.split('\n');

  const minIndent = lines
    .filter(line => line.trim())
    .reduce<number>((minIndent, line) => {
      const currIndent = line.match(/^ */)?.[0].length;
      if (currIndent == null) {
        return minIndent;
      }

      return currIndent < minIndent ? currIndent : minIndent;
    }, Infinity);

  return lines
    .map(line => line.slice(minIndent))
    .join('\n')
    .trim()
    .replace(/\n +\n/g, '\n\n');
};

// foo_barBaz -> ['foo', 'bar', 'Baz']
export const splitWords = (str: string) => {
  return str.replace(/[A-Z]/, ' $&').split(/[^a-zA-Z0-9]+/);
};

// upper -> Upper
export const toUpperFirst = (str: string) => {
  return str.slice(0, 1).toUpperCase() + str.slice(1).toLowerCase();
};

// foo-bar-baz -> fooBarBaz
export const toCamelCase = (str: string) => {
  const words = splitWords(str);
  const first = words.shift()?.toLowerCase() ?? '';
  const rest = words.map(toUpperFirst);

  return [first, ...rest].join('');
};
