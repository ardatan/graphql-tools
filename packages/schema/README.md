Check API Reference for more information about this package;
https://www.graphql-tools.com/docs/api/modules/schema_src

You can also learn more about Generating Executable Schemas in this chapter;
https://www.graphql-tools.com/docs/generate-schema

