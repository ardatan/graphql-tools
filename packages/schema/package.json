{
  "name": "@graphql-tools/schema",
  "version": "8.3.12",
  "description": "A set of utils for faster development of GraphQL tools",
  "repository": {
    "type": "git",
    "url": "ardatan/graphql-tools",
    "directory": "packages/schema"
  },
  "license": "MIT",
  "sideEffects": false,
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "exports": {
    ".": {
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./*": {
      "require": "./dist/*.js",
      "import": "./dist/*.mjs"
    }
  },
  "typings": "dist/index.d.ts",
  "typescript": {
    "definition": "dist/index.d.ts"
  },
  "peerDependencies": {
    "graphql": "^14.0.0 || ^15.0.0 || ^16.0.0"
  },
  "devDependencies": {
    "graphql-scalars": "1.17.0"
  },
  "buildOptions": {
    "input": "./src/index.ts"
  },
  "dependencies": {
    "@graphql-tools/merge": "8.2.12",
    "@graphql-tools/utils": "8.6.11",
    "tslib": "~2.4.0",
    "value-or-promise": "1.0.11"
  },
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  }
}
