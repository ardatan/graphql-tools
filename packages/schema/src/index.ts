export { assertResolversPresent } from './assertResolversPresent.js';
export { chainResolvers } from './chainResolvers.js';
export { addResolversToSchema } from './addResolversToSchema.js';
export { checkForResolveTypeResolver } from './checkForResolveTypeResolver.js';
export { extendResolversFromInterfaces } from './extendResolversFromInterfaces.js';
export * from './makeExecutableSchema.js';
export * from './types.js';
export * from './merge-schemas.js';
export { extractExtensionsFromSchema } from '@graphql-tools/utils';
