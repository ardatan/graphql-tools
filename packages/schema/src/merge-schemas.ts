import { GraphQLSchema } from 'graphql';
import {
  IResolvers,
  asArray,
  getResolversFromSchema,
  TypeSource,
  SchemaExtensions,
  extractExtensionsFromSchema,
} from '@graphql-tools/utils';
import { makeExecutableSchema } from './makeExecutableSchema.js';
import { IExecutableSchemaDefinition } from './types.js';

/**
 * Configuration object for schema merging
 */
export type MergeSchemasConfig<T = any> = Partial<IExecutableSchemaDefinition<T>> & {
  /**
   * The schemas to be merged
   */
  schemas?: GraphQLSchema[];
};

/**
 * Synchronously merges multiple schemas, typeDefinitions and/or resolvers into a single schema.
 * @param config Configuration object
 */
export function mergeSchemas(config: MergeSchemasConfig) {
  const extractedTypeDefs: TypeSource[] = [];
  const extractedResolvers: IResolvers<any, any>[] = [];
  const extractedSchemaExtensions: SchemaExtensions[] = [];

  if (config.schemas != null) {
    for (const schema of config.schemas) {
      extractedTypeDefs.push(schema);
      extractedResolvers.push(getResolversFromSchema(schema));
      extractedSchemaExtensions.push(extractExtensionsFromSchema(schema));
    }
  }

  if (config.typeDefs != null) {
    extractedTypeDefs.push(config.typeDefs);
  }

  if (config.resolvers != null) {
    const additionalResolvers = asArray(config.resolvers);
    extractedResolvers.push(...additionalResolvers);
  }

  if (config.schemaExtensions != null) {
    const additionalSchemaExtensions = asArray(config.schemaExtensions);
    extractedSchemaExtensions.push(...additionalSchemaExtensions);
  }

  return makeExecutableSchema({
    ...config,
    typeDefs: extractedTypeDefs,
    resolvers: extractedResolvers,
    schemaExtensions: extractedSchemaExtensions,
  });
}
