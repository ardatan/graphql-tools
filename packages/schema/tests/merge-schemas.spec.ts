import { makeExecutableSchema } from '@graphql-tools/schema';
import { graphql, buildSchema, GraphQLScalarType, Kind, GraphQLSchema, print } from 'graphql';
import { mergeSchemas } from '../src/merge-schemas.js';
import { assertSome, printSchemaWithDirectives } from '@graphql-tools/utils';
import { assertListValueNode } from '../../testing/assertion.js';

describe('Merge Schemas', () => {
  it('Should include extensions in merged schemas', () => {
    const fooSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          foo: String
        }
      `,
      resolvers: {
        Query: {
          foo: () => 'FOO',
        },
      },
    });
    fooSchema.extensions = { schemaA: true };
    const barSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          bar: String
        }
      `,
      resolvers: {
        Query: {
          bar: () => 'BAR',
        },
      },
    });
    barSchema.extensions = { schemaB: true };

    const mergedSchema = mergeSchemas({
      schemas: [fooSchema, barSchema],
    });

    expect(mergedSchema.extensions).toEqual({ schemaA: true, schemaB: true });
  });

  it('should override resolver in schema with resolver passed into config', async () => {
    const fooSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          foo: String
        }
      `,
      resolvers: {
        Query: {
          foo: () => 'FOO',
        },
      },
    });
    const barSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          bar: String
        }
      `,
      resolvers: {
        Query: {
          bar: () => 'BAR',
        },
      },
    });
    const { errors, data } = await graphql({
      schema: mergeSchemas({
        schemas: [fooSchema, barSchema],
        typeDefs: /* GraphQL */ `
          type Query {
            qux: String
          }
        `,
        resolvers: {
          Query: {
            qux: () => 'QUX',
            foo: () => 'FOO_BAR_QUX',
          },
        },
      }),
      source: `
                {
                    foo
                    bar
                    qux
                }
            `,
    });
    expect(errors).toBeFalsy();
    assertSome(data);
    expect(data['foo']).toBe('FOO_BAR_QUX');
    expect(data['bar']).toBe('BAR');
    expect(data['qux']).toBe('QUX');
  });

  it('should merge two valid executable schemas', async () => {
    const fooSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          foo: String
        }
      `,
      resolvers: {
        Query: {
          foo: () => 'FOO',
        },
      },
    });
    const barSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          bar: String
        }
      `,
      resolvers: {
        Query: {
          bar: () => 'BAR',
        },
      },
    });
    const { errors, data } = await graphql({
      schema: mergeSchemas({
        schemas: [fooSchema, barSchema],
      }),
      source: `
                {
                    foo
                    bar
                }
            `,
    });
    expect(errors).toBeFalsy();
    assertSome(data);
    expect(data['foo']).toBe('FOO');
    expect(data['bar']).toBe('BAR');
  });
  it('should merge two valid executable schemas with extra resolvers', async () => {
    const fooSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          foo: String
        }
      `,
      resolvers: {
        Query: {
          foo: () => 'FOO',
        },
      },
    });
    const barSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          bar: String
          qux: String
        }
      `,
      resolvers: {
        Query: {
          bar: () => 'BAR',
        },
      },
    });
    const { errors, data } = await graphql({
      schema: mergeSchemas({
        schemas: [fooSchema, barSchema],
        resolvers: {
          Query: {
            qux: () => 'QUX',
          },
        },
      }),
      source: `
                {
                    foo
                    bar
                    qux
                }
            `,
    });
    expect(errors).toBeFalsy();
    assertSome(data);
    expect(data['foo']).toBe('FOO');
    expect(data['bar']).toBe('BAR');
    expect(data['qux']).toBe('QUX');
  });
  it('should merge two valid executable schemas with extra typeDefs and resolvers', async () => {
    const fooSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          foo: String
        }
      `,
      resolvers: {
        Query: {
          foo: () => 'FOO',
        },
      },
    });
    const barSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          bar: String
        }
      `,
      resolvers: {
        Query: {
          bar: () => 'BAR',
        },
      },
    });
    const { errors, data } = await graphql({
      schema: mergeSchemas({
        schemas: [fooSchema, barSchema],
        typeDefs: /* GraphQL */ `
          type Query {
            qux: String
          }
        `,
        resolvers: {
          Query: {
            qux: () => 'QUX',
          },
        },
      }),
      source: `
                {
                    foo
                    bar
                    qux
                }
            `,
    });
    expect(errors).toBeFalsy();
    assertSome(data);
    expect(data['foo']).toBe('FOO');
    expect(data['bar']).toBe('BAR');
    expect(data['qux']).toBe('QUX');
  });
  it('should merge two valid schemas by keeping their directives to be used in extra typeDefs', async () => {
    const fooSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        directive @fooDirective on FIELD_DEFINITION
        type Query {
          foo: String
        }
      `,
      resolvers: {
        Query: {
          foo: () => 'FOO',
        },
      },
    });
    const barSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          bar: String
        }
      `,
      resolvers: {
        Query: {
          bar: () => 'BAR',
        },
      },
    });
    const { errors, data } = await graphql({
      schema: mergeSchemas({
        schemas: [fooSchema, barSchema],
        typeDefs: /* GraphQL */ `
          type Query {
            qux: String @fooDirective
          }
        `,
        resolvers: {
          Query: {
            qux: () => 'QUX',
          },
        },
      }),
      source: `
                {
                    foo
                    bar
                    qux
                }
            `,
    });
    expect(errors).toBeFalsy();
    assertSome(data);
    expect(data['foo']).toBe('FOO');
    expect(data['bar']).toBe('BAR');
    expect(data['qux']).toBe('QUX');
  });
  it('should merge valid schemas with interfaces correctly', async () => {
    const fooSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        interface Foo {
          foo: String
        }
        type Bar implements Foo {
          foo: String
          bar: String
        }
        type Qux implements Foo {
          foo: String
          qux: String
        }
      `,
    });
    const barSchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        interface Foo {
          foo: String
        }
        type Query {
          bar: Foo
          qux: Foo
        }
      `,
      resolvers: {
        Foo: {
          __resolveType: (root: any) => {
            if ('bar' in root) {
              return 'Bar';
            }
            if ('qux' in root) {
              return 'Qux';
            }
            return null;
          },
        },
        Query: {
          bar: () => ({ foo: 'foo', bar: 'bar' }),
          qux: () => ({ foo: 'foo', qux: 'qux' }),
        },
      },
    });
    const result = await graphql({
      schema: mergeSchemas({
        schemas: [fooSchema, barSchema],
      }),
      source: /* GraphQL */ `
        {
          bar {
            foo
            ... on Bar {
              bar
            }
          }
          qux {
            foo
            ... on Qux {
              qux
            }
          }
        }
      `,
    });
    const { data, errors } = result as any;
    expect(errors).toBeFalsy();
    assertSome(data);
    expect(data['bar'].foo).toBe('foo');
    expect(data['bar'].bar).toBe('bar');
    expect(data['qux'].foo).toBe('foo');
    expect(data['qux'].qux).toBe('qux');
  });

  it('should merge scalars (part of resolvers)', async () => {
    const now = new Date();
    const schemaA = makeExecutableSchema({
      typeDefs: [`scalar Date`, `type Query { a: Date }`],
      resolvers: {
        Query: {
          a: () => now,
        },
        Date: new GraphQLScalarType({
          name: 'Date',
          serialize(value) {
            if (typeof value === 'string' || typeof value === 'number' || value instanceof Date) {
              return new Date(value).toISOString();
            }
          },
          parseValue(value) {
            if (typeof value === 'string' || typeof value === 'number' || value instanceof Date) {
              return new Date(value);
            }
          },
          parseLiteral(ast) {
            if (ast.kind !== Kind.STRING) {
              throw new TypeError(`Date cannot represent non string type`);
            }
            return new Date(ast.value);
          },
        }),
      },
    });
    const schemaB = makeExecutableSchema({
      typeDefs: [`type Query { b: String }`],
    });

    const schema = mergeSchemas({ schemas: [schemaA, schemaB] });

    // original schema A
    const { data: dataA } = await graphql({
      schema: schemaA,
      source: /* GraphQL */ `
        {
          a
        }
      `,
    });

    assertSome(dataA);
    expect(dataA['a']).toEqual(now.toISOString());

    // merged schema
    const { data } = await graphql({
      schema,
      source: /* GraphQL */ `
        {
          a
        }
      `,
    });
    assertSome(data);
    expect(data['a']).toEqual(now.toISOString());
  });

  it('should not duplicate directives of scalars', () => {
    const schema = buildSchema(/* GraphQL */ `
      directive @sqlType(type: String!) on SCALAR
      scalar JSON @sqlType(type: "json")
    `);
    const merged = mergeSchemas({ schemas: [schema] });
    expect(print(merged.getType('JSON')!.astNode!)).toEqual('scalar JSON @sqlType(type: "json")');
  });

  it('should merge when directive uses enum', () => {
    const merged = mergeSchemas({
      schemas: [
        buildSchema(/* GraphQL */ `
          directive @date(format: DateFormat) on FIELD_DEFINITION

          enum DateFormat {
            LOCAL
            ISO
          }
        `),
      ],
      typeDefs: /* GraphQL */ `
        scalar Date
        type Query {
          today: Date @date
        }
      `,
    });

    expect(merged.getDirective('date')).toBeDefined();
  });
  it('should merge a lot of directives but without high memory usage', () => {
    let num = 100;
    const base = buildSchema(/* GraphQL */ `
      directive @access(roles: [String]) on FIELD_DEFINITION

      type Query {
        test: Boolean @access(roles: ["Admin"])
      }
    `);

    let prev: GraphQLSchema = base;

    while (num--) {
      prev = mergeSchemas({ schemas: [prev, base] });
    }
    const QueryType = prev.getQueryType();
    assertSome(QueryType);
    const fields = QueryType.getFields();
    assertSome(fields['test'].astNode);
    assertSome(fields['test'].astNode.directives);
    assertSome(fields['test'].astNode.directives[0]);
    assertSome(fields['test'].astNode.directives[0].arguments);
    assertListValueNode(fields['test'].astNode.directives[0].arguments[0].value);

    expect(fields['test'].astNode.directives[0].arguments[0].value.values).toHaveLength(1);
  });
  it('should merge schemas with custom scalars', () => {
    const GraphQLUUID = new GraphQLScalarType({
      name: 'UUID',
      serialize: val => val,
      parseValue: val => val,
      parseLiteral: valueNode => {
        if (valueNode.kind === 'StringValue') {
          return valueNode.value;
        }
        return null;
      },
    });
    const countrySchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        scalar UUID
        type Country {
          id: UUID!
          name: String
        }
        type Query {
          country: Country
        }
      `,
      resolvers: {
        UUID: GraphQLUUID,
      },
    });

    const citySchema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        scalar UUID
        type City {
          id: ID!
          name: String
        }
        type Query {
          city: City
        }
      `,
    });

    const mergedSchema = mergeSchemas({
      schemas: [countrySchema, citySchema],
    });
    const printedSchema = printSchemaWithDirectives(mergedSchema);

    expect(printedSchema).toContain(
      /* GraphQL */ `
type Query {
  country: Country
  city: City
}`.trim()
    );
    expect(printedSchema).toContain(
      /* GraphQL */ `
scalar UUID
`.trim()
    );
    expect(printedSchema).toContain(
      /* GraphQL */ `
type Country {
  id: UUID!
  name: String
}`.trim()
    );
    expect(printedSchema).toContain(
      /* GraphQL */ `
type City {
  id: ID!
  name: String
}`.trim()
    );
  });
});
