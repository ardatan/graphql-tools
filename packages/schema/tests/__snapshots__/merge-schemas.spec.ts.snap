// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merge Schemas should merge schemas with directives in extensions 1`] = `
"schema @onSchema(name: "a") @onSchema(name: "b") {
  query: Query
}

type Query @onType(name: "a") @onType(name: "b") {
  a: String @onField(name: "a") @onField(name: "b")
}"
`;
